lib/net.c

```diff
@@ -42,17 +44,45 @@
 
 static struct net_config config;
 
-/*
- * XXX The secret key of the RSS hash must be random
- * in order to avoid hackers to know it.
- */
-uint8_t default_rss_key[GATEKEEPER_RSS_KEY_LEN] = {
-	0x6d, 0x5a, 0x56, 0xda, 0x25, 0x5b, 0x0e, 0xc2,
-	0x41, 0x67, 0x25, 0x3d, 0x43, 0xa3, 0x8f, 0xb0,
-	0xd0, 0xca, 0x2b, 0xcb, 0xae, 0x7b, 0x30, 0xb4,
-	0x77, 0xcb, 0x2d, 0xa3, 0x80, 0x30, 0xf2, 0x0c,
-	0x6a, 0x42, 0xb7, 0x3b, 0xbe, 0xac, 0x01, 0xfa,
-};
+uint8_t default_rss_key[GATEKEEPER_RSS_KEY_LEN];
+
+static void
+randomize_rss_key(void)
+{
+	uint8_t i;
+	uint16_t final_set_count = 0;
+	ssize_t number_of_bytes;
+
+	/*
+	 * The secret key of the RSS hash (RSK) must be random in order
+	 * to prevent hackers from knowing it.
+	 * To validate if the key generated is reasonable, the
+	 * number of bits set to 1 in the key must be greater than 
+	 * 10% and less than 90% of the total bits in the key.
+	 * min_num_set_bits and max_num_set_bits represent the lower 
+	 * and upper bound for the key.
+	 */
+	uint16_t min_num_set_bits = sizeof(default_rss_key) * 8 * 0.1;
+	uint16_t max_num_set_bits = sizeof(default_rss_key) * 8 * 0.9;
+
+	while (final_set_count < min_num_set_bits ||
+			final_set_count > max_num_set_bits) {
+		final_set_count = 0;
+		/* 
+		 * The last parameter to syscall getrandom i.e flags will alter the 
+		 * behaviour of the call. In the case where flag == 0, the call will
+		 * block until the pool has the required entropy.  The GRND_RANDOM flag
+		 * bit can be used to switch to the /dev/random pool.
+		 */
+		while ((number_of_bytes = syscall(SYS_getrandom, default_rss_key, sizeof(default_rss_key), 0)) !=
```

Comment:
If `getrandom()` returns an error, this code will loop forever.

Reply:
I think it would be better to avoid the while loop at all. It returns the total number of bytes on success, which can be used to decide if it executed successfully or -1 on error. I'd vote for a recursive call - if the number of bytes are less than the size, it can be called again recursively. Or it can exit if there's an error. However, the probability of such a case is very less as it is in blocking mode(it returns only after it has sufficient entropy thus filling all the bytes) using flag `0` passed to it

Reply:
Waiting for your input on this

Reply:
First of all, there's no need for relying on probabilistic code here since writing robust code is straightforward.

Second, while recursions are often elegant, one must know its tools well. By definition, the C language does not support tail call, that is, C favors loops over recursions. It's not to say that recursions are forbidden, and loops are always best. The case here asks for a straightforward `do  { ... getrandom() ... if (UNRECOVERABLE ERROR) ... } while (RECOVERABLE ERROR);`.

Finally, the code can even be better. A variable can accumulate the number of random bytes read, and whenever less than needed bytes are read (a recoverable error), the following call of `getrandom()` can use that information to shift the first parameter and decrement the second to only request the remaining bytes.